include/CL/sycl/buffer/detail/buffer.hpp

```diff
@@ -119,11 +143,52 @@ class buffer : public detail::buffer_base,
       used.
   */
   buffer(shared_ptr_class<T> &host_data,
-         const range<Dimensions> &r)
-    : buffer_base { false },
+         const range<Dimensions> &r) :
+    buffer_base { false },
     access { host_data.get(), r },
-    shared_data { host_data }
-    {}
+    shared_pointer { host_data },
+    data_host { true }
+  {}
+
+
+  /** Create a new buffer with associated memory, using the data in
+   *  host_data.
+   *
+   *  SYCL's runtime has full ownership of the host_data.
+   */
+  buffer(unique_ptr_class<T> &&host_data,
+         const range<Dimensions> &r) :
+    buffer_base { false },
+    access { host_data.get(), r },
+    unique_pointer { host_data },
+    data_host { true }
+  {}
+
+
+private:
+
+
+  /* On these two methods are called by the constructor when used with two iterators.
```

Comment:
s/On/One of/ ?
I do not really understand the name "constructor_for_iterator".
Furthermore "Mode" is unclear.
Remove the *